src/server/middlewares/validateVote.js

```diff
@@ -1,175 +1,260 @@
+
+      const result = await questionContract.methods
+        .getAnswer(answerChainIds[index])
+        .call();
+      const isInitialized = result[0];
+
+      if (!isInitialized) {
+        throw Error('Invalid answer');
       }
     }),
   );
-  if (sum > maxVoteTokens * answers.length) {
-    throw Error('too many votes');
+
+  if (sum > maxVoteTokens) {
+    throw Error('Too many votes');
   }
-};
+}
 
-const checkAnswers = async (vote, festivalQuestion, artworkQuestion) => {
-  if (vote.festivalAnswers.length !== vote.festivalVoteTokens.length) {
-    throw Error('wrong array length');
+async function checkAnswers(
+  vote,
+  festivalQuestionContract,
+  artworkQuestionContract,
+) {
+  if (vote.festivalAnswerIds.length !== vote.festivalVoteTokens.length) {
+    throw Error('Wrong festival answer length');
   }
-  if (vote.artworkAnswers.length !== vote.artworkVoteTokens.length) {
-    throw Error('wrong array length');
+
+  if (vote.artworkAnswerIds.length !== vote.artworkVoteTokens.length) {
+    throw Error('Wrong artwork answer length');
   }
+
   await checkMaxVotes(
-    vote.festivalAnswers,
-    vote.festivalVotes,
-    festivalQuestion,
+    vote.festivalAnswerIds,
+    vote.festivalAnswerChainIds,
+    vote.festivalVoteTokens,
+    festivalQuestionContract,
   );
-  await checkMaxVotes(vote.artworkAnswers, vote.artworkVotes, artworkQuestion);
-};
 
-export default async function (req, res, next) {
-  const vote = req.body;
-  const festivalQuestion = getQuestionContract(vote.festivalQuestion);
-  const artworkQuestion = getQuestionContract(vote.artworkQuestion);
+  await checkMaxVotes(
+    vote.artworkAnswerIds,
+    vote.artworkAnswerChainIds,
+    vote.artworkVoteTokens,
+    artworkQuestionContract,
+  );
+}
+
+export default async function validateVoteMiddleware(req, res, next) {
+  const { vote } = req.locals;
+
+  const festivalContract = getQuestionContract(vote.festivalQuestionAddress);
+  const artworkContract = getQuestionContract(vote.artworkQuestionAddress);
+
   try {
-    await checkSigs(vote);
-    await checkBooth(vote, festivalQuestion);
-    await checkBooth(vote, artworkQuestion);
+    await checkSignatures(vote);
+    await checkBooth(vote, festivalContract);
+    await checkBooth(vote, artworkContract);
     await checkNonce(vote);
-    await checkChainHasVoted(vote, festivalQuestion, artworkQuestion);
+    await checkHasVoted(vote, festivalContract, artworkContract);
     await checkLocalHasVoted(vote);
     await checkQuestions(vote);
-    await checkFestival(vote);
+    await checkFestival(vote, festivalContract);
     await checkArtworkQuestionIsHighestVoted(vote);
-    await checkAnswers(vote, festivalQuestion, artworkQuestion);
+    await checkAnswers(vote, festivalContract, artworkContract);
+
     next();
-  } catch (err) {
-    next(new APIError(httpStatus.BAD_REQUEST));
+  } catch (error) {
+    logger.verbose(error);
+    next(new APIError(httpStatus.UNPROCESSABLE_ENTITY, error.message));
```

Comment:
Fwiw, the rationale for using a generic error was the security best practice of not returning helpful information to a possible attacker